### ISSUE TITLE
diffractometer: add pa() report

### ISSUE THREAD

**Issue by prjemian** [BCDA-APS/apstools#434]
Diffractometer needs a `pa()`-style report that shows all of what it knows.  Similar to the SPEC `pa` command or this image:

![Clipboard01](https://user-images.githubusercontent.com/2279984/94486305-d6816280-01a4-11eb-94e4-f97977c70141.png)

**Comment by prjemian** [comment]
Like this:

```
In [9]: sim4c.pa()                                                                                                                                               
===================== ===============================================================================
term                  value                                                                          
===================== ===============================================================================
diffractometer        sim4c                                                                          
geometry              E4CV                                                                           
class                 MyE4CV                                                                         
energy (keV)          0.80509                                                                        
wavelength (angstrom) 1.54000                                                                        
calc engine           hkl                                                                            
mode                  bissector                                                                      
positions             ======= ========= =============                                                
                      name    value     original name                                                
                      ======= ========= =============                                                
                      able    -45.00000 omega                                                        
                      baker   135.00000 chi                                                          
                      charlie -90.00000 phi                                                          
                      arm     -90.00000 tth                                                          
                      ======= ========= =============                                                
sample: main          ================= =============================================================
                      term              value                                                        
                      ================= =============================================================
                      unit cell edges   a=1.54, b=1.54, c=1.54                                       
                      unit cell angles  alpha=90.0, beta=90.0, gamma=90.0                            
                      ref 1 (hkl)       h=1.0, k=1.0, l=0.0                                          
                      ref 1 positioners able=45.00000, baker=45.00000, charlie=90.00000, arm=90.00000
                      ref 2 (hkl)       h=0.0, k=1.0, l=0.0                                          
                      ref 2 positioners able=30.00000, baker=90.00000, charlie=90.00000, arm=60.00000
                      [U]               [[ 1.00000000e+00 -1.11022302e-16  1.08845649e-16]           
                                         [ 1.11022302e-16  1.00000000e+00 -1.08845649e-16]           
                                         [-1.08845649e-16  1.08845649e-16  1.00000000e+00]]          
                      [UB]              [[ 4.07999046e+00 -7.02797298e-16  1.94261847e-16]           
                                         [ 4.52969935e-16  4.07999046e+00 -6.93916573e-16]           
                                         [-4.44089210e-16  4.44089210e-16  4.07999046e+00]]          
                      ================= =============================================================
===================== ===============================================================================

Out[9]: <pyRestTable.rest_table.Table at 0x7effdea71be0>
```

**Comment by prjemian** [comment]
Now, with constraints reported:

```python
(bluesky_2020_9) prjemian@poof ~/.../projects/apstools $ ipython
Python 3.8.2 (default, Mar 26 2020, 15:53:00) 
Type 'copyright', 'credits' or 'license' for more information
IPython 7.16.1 -- An enhanced Interactive Python. Type '?' for help.

   ...: import gi
   ...: gi.require_version('Hkl', '5.0')    # MUST come before `import hkl` 
   ...: from apstools import diffractometer as APS_diffractometer 
   ...: from ophyd import Component, Device, PseudoSingle, SoftPositioner 
   ...: import hkl.diffract 
   ...:  
   ...: class MyE4CV(APS_diffractometer.DiffractometerMixin, hkl.diffract.E4CV): 
   ...:  
   ...:     able = Component(SoftPositioner, 
   ...:         labels=("motor", "sim4c"), kind="hinted") 
   ...:     baker =   Component(SoftPositioner, 
   ...:         labels=("motor", "sim4c"), kind="hinted") 
   ...:     charlie =   Component(SoftPositioner, 
   ...:         labels=("motor", "sim4c"), kind="hinted") 
   ...:     arm =   Component(SoftPositioner, 
   ...:         labels=("motor", "sim4c"), kind="hinted") 
   ...:  
   ...:     def __init__(self, *args, **kwargs): 
   ...:         super().__init__(*args, **kwargs) 
   ...:         for axis in self.real_positioners: 
   ...:             axis.move(0) 
   ...:  
   ...: sim4c = MyE4CV('', name='sim4c') 
   ...: sim4c.calc.physical_axis_names = dict( 
   ...:     omega = "able", 
   ...:     chi = "baker", 
   ...:     phi = "charlie", 
   ...:     tth = "arm") 
   ...:  
   ...: rp1 = sim4c.calc.Position(able=45, baker=45, charlie=90.0, arm=90) 
   ...: rp2 = sim4c.calc.Position(able=30, baker=90, charlie=90.0, arm=60) 
   ...: r1 = sim4c.calc.sample.add_reflection(1, 1, 0, position=rp1) 
   ...: r2 = sim4c.calc.sample.add_reflection(0, 1, 0, position=rp2) 
   ...: sim4c.calc.sample.compute_UB(r1, r2) 
   ...:                                                                                                                                                          
Out[1]: 1

In [2]: sim4c.pa()                                                                                                                                               
===================== ===============================================================================
term                  value                                                                          
===================== ===============================================================================
diffractometer        sim4c                                                                          
geometry              E4CV                                                                           
class                 MyE4CV                                                                         
energy (keV)          0.80509                                                                        
wavelength (angstrom) 1.54000                                                                        
calc engine           hkl                                                                            
mode                  bissector                                                                      
positions             ======= ======= =============                                                  
                      name    value   original name                                                  
                      ======= ======= =============                                                  
                      able    0.00000 omega                                                          
                      baker   0.00000 chi                                                            
                      charlie 0.00000 phi                                                            
                      arm     0.00000 tth                                                            
                      ======= ======= =============                                                  
constraints           ======= ========= ========== ===== ====                                        
                      axis    low_limit high_limit value fit                                         
                      ======= ========= ========== ===== ====                                        
                      able    -180.0    180.0      0.0   True                                        
                      baker   -180.0    180.0      0.0   True                                        
                      charlie -180.0    180.0      0.0   True                                        
                      arm     -180.0    180.0      0.0   True                                        
                      ======= ========= ========== ===== ====                                        
sample: main          ================= =============================================================
                      term              value                                                        
                      ================= =============================================================
                      unit cell edges   a=1.54, b=1.54, c=1.54                                       
                      unit cell angles  alpha=90.0, beta=90.0, gamma=90.0                            
                      ref 1 (hkl)       h=1.0, k=1.0, l=0.0                                          
                      ref 1 positioners able=45.00000, baker=45.00000, charlie=90.00000, arm=90.00000
                      ref 2 (hkl)       h=0.0, k=1.0, l=0.0                                          
                      ref 2 positioners able=30.00000, baker=90.00000, charlie=90.00000, arm=60.00000
                      [U]               [[ 1.00000000e+00 -1.11022302e-16  1.08845649e-16]           
                                         [ 1.11022302e-16  1.00000000e+00 -1.08845649e-16]           
                                         [-1.08845649e-16  1.08845649e-16  1.00000000e+00]]          
                      [UB]              [[ 4.07999046e+00 -7.02797298e-16  1.94261847e-16]           
                                         [ 4.52969935e-16  4.07999046e+00 -6.93916573e-16]           
                                         [-4.44089210e-16  4.44089210e-16  4.07999046e+00]]          
                      ================= =============================================================
===================== ===============================================================================

Out[2]: <pyRestTable.rest_table.Table at 0x7fc9e3035430>
```

**Comment by prjemian** [comment]
Here's the code that sets up the test of the `pa()` and `wa()` methods.  It also shows how to use custom motor names:

```python
import gi
gi.require_version('Hkl', '5.0')    # MUST come before `import hkl`
from apstools import diffractometer as APS_diffractometer
from ophyd import Component, Device, PseudoSingle, SoftPositioner
import hkl.diffract

class MyE4CV(APS_diffractometer.DiffractometerMixin, hkl.diffract.E4CV):

    able = Component(SoftPositioner,
        labels=("motor", "sim4c"), kind="hinted")
    baker =   Component(SoftPositioner,
        labels=("motor", "sim4c"), kind="hinted")
    charlie =   Component(SoftPositioner,
        labels=("motor", "sim4c"), kind="hinted")
    arm =   Component(SoftPositioner,
        labels=("motor", "sim4c"), kind="hinted")

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        for axis in self.real_positioners:
            axis.move(0)

sim4c = MyE4CV('', name='sim4c')
sim4c.calc.physical_axis_names = dict(
    omega = "able",
    chi = "baker",
    phi = "charlie",
    tth = "arm")

rp1 = sim4c.calc.Position(able=45, baker=45, charlie=90.0, arm=90)
rp2 = sim4c.calc.Position(able=30, baker=90, charlie=90.0, arm=60)
r1 = sim4c.calc.sample.add_reflection(1, 1, 0, position=rp1)
r2 = sim4c.calc.sample.add_reflection(0, 1, 0, position=rp2)
sim4c.calc.sample.compute_UB(r1, r2)
```

**Comment by gfabbris** [comment]
Looks good! 

One thing that is different from spec's pa is that the wavelength at which the r1/r2 were collected is not reported here. I briefly looked into this and it seems like the reflection does not hold the energy at which the angles were measured, am I correct? When we `compute_UB` does it assume that the reflections were measured at the current energy?

**Comment by prjemian** [comment]
Great catch!  Yes, the `compute_UB()` method assumes the reflections are at the same wavelength which is the current one.  Wonder what the C++ code assumes?  Probably the same.

Needs a new issue to record all this together.  (Had to jump through hoops to get the angles associated with each _hkl_ reflection.)  A concise record would clarify this.

**Comment by prjemian** [comment]
Once the issue in hklpy is complete, `pa()` can be refactored to include wavelength.
